### PR TITLE
Append stack trace to fx node

### PR DESCRIPTION
**Pull request by SherlockNoMad**
For debugging purpose, user commonly wish to trace back from an fx.Node to the user code. 

image
Sample output: 

```
placeholder x
  File "/fsx/users/bahuang/repos/torchdynamo/test.py", line 21, in forward
    def forward(self, x):


call_function <built-in method sqrt of type object at 0x7f4c3f8d8360>
  File "/fsx/users/bahuang/repos/torchdynamo/test.py", line 22, in forward
    x = self.n(x)
  File "/fsx/users/bahuang/repos/torchdynamo/test.py", line 12, in forward
    a = torch.sqrt(q)


call_method relu
  File "/fsx/users/bahuang/repos/torchdynamo/test.py", line 22, in forward
    x = self.n(x)
  File "/fsx/users/bahuang/repos/torchdynamo/test.py", line 13, in forward
    return f(a)
  File "/fsx/users/bahuang/repos/torchdynamo/test.py", line 8, in f
    return x.relu()


call_module self_sigmoid
  File "/fsx/users/bahuang/repos/torchdynamo/test.py", line 23, in forward
    x = self.sigmoid(x)


output output
None
```

### CODE REVIEW

File: torchdynamo/output_graph.py [hunk 1]
```diff
@@ -433,3 +433,12 @@ def cleanup(self):
         for node in self.graph.nodes:
             if "example_value" in node.meta:
                 del node.meta["example_value"]
+
+    def create_proxy(self, kind, target, args, kwargs, name=None, type_expr=None, proxy_factory_fn=None):
+        rv = super().create_proxy(kind, target, args, kwargs, name, type_expr, proxy_factory_fn)
+
+        # append stack trace to fx node
```
Comment (SherlockNoMad) [inline torchdynamo/output_graph.py:440]: Do we need some debug flag to enable this？

File: torchdynamo/output_graph.py [hunk 2]
```diff
@@ -433,3 +433,12 @@ def cleanup(self):
...
+        frame_summary = self.root_tx.frame_summary()
```
Comment (SherlockNoMad) [inline torchdynamo/output_graph.py:441]: hmm... using self.root_tx doesn't handle nested functions. 

I still need to handle nested InliningInstructionTranslator ...

**Comment by voznesenskym** [comment]
J/w Why do we want to append stack trace to fx nodes?

**Comment by SherlockNoMad** [comment]
> J/w Why do we want to append stack trace to fx nodes?

For debugging purpose... it's a common request to correlate back to the user code from an FX node.

File: torchdynamo/output_graph.py (hunk 2 above)
Comment (SherlockNoMad) [reply torchdynamo/output_graph.py]: fixed.

File: torchdynamo/output_graph.py (hunk 1 above)
Comment (jansel) [reply torchdynamo/output_graph.py:452]: I'm assuming this track traces will be small, so I don't see a strong need.

File: torchdynamo/output_graph.py [hunk 3]
```diff
@@ -433,3 +433,25 @@ def cleanup(self):
...
+        current_tx=None,
+    ):
+        rv = super().create_proxy(
+            kind, target, args, kwargs, name, type_expr, proxy_factory_fn
+        )
+
+        # append stack trace to fx node
+        tx = current_tx if current_tx else self.root_tx
+        frame_summary = tx.frame_summary()
+        rv.node.stack_trace = f"{frame_summary}\n  {frame_summary.line}"
```
Comment (jansel) [inline torchdynamo/output_graph.py:455]: If you want to print a `traceback.FrameSummary` you could use code like:
https://github.com/pytorch/torchdynamo/blob/9ecbdaed5c1afe2f8443e0f13e081191cc29b0d3/torchdynamo/exc.py#L55
which I'd expect to look a bit nicer.

The other option would be to leave it as a `traceback.FrameSummary` object, which will allow users to query file/line more easily.

File: torchdynamo/output_graph.py [hunk 4]
```diff
@@ -433,3 +433,25 @@ def cleanup(self):
...
+    def create_proxy(
+        self,
+        kind,
+        target,
+        args,
+        kwargs,
+        name=None,
+        type_expr=None,
+        proxy_factory_fn=None,
+        current_tx=None,
```
Comment (anijain2305) [inline torchdynamo/output_graph.py:446]: Soft suggestion - feel free to ignore - replace `current_tx` to just `tx`?

**Review by anijain2305 — COMMENTED** [review]
Thanks! Glad that the utility was already there. It turned out to be exactly what we needed :)

File: torchdynamo/output_graph.py (hunk 1 above)
Comment (SherlockNoMad) [reply torchdynamo/output_graph.py:452]: ok.

File: torchdynamo/output_graph.py (hunk 4 above)
Comment (SherlockNoMad) [reply torchdynamo/output_graph.py:446]: I will keep it as current_tx, as it's a stack of txs....

File: torchdynamo/output_graph.py (hunk 3 above)
Comment (SherlockNoMad) [reply torchdynamo/output_graph.py]: done.

**Review by anijain2305 — APPROVED** [review]

**Pull request merged**